### PR TITLE
Allow predict() to potentially return RaggedTensor

### PR DESCRIPTION
**Pull request by Rocketknight1**
Currently, predict() throws an error if the batches it's concatenating have variable sizes. This is a headache in NLP in particular, because batches with variable sequence length dimensions are very common. This change maintains the current behaviour when batches can be concatenated into a dense Tensor, but when they can't it attempts to return a RaggedTensor instead, while inferring as many constant inner dimensions as it can.

I discussed this with @fchollet already and he approved in principle, but please let me know if there are any issues with my implementation - I know `tf.ragged.constant` in particular can have a very long runtime if you're not careful with the inputs, but I believe it's efficient when the inputs are Numpy arrays.

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -2011,7 +2011,7 @@ def predict(self,
                          'information of where went wrong, or file a '
                          'issue/bug to `tf.keras`.')
       callbacks.on_predict_end()
-    all_outputs = tf.__internal__.nest.map_structure_up_to(batch_outputs, concat, outputs)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:2014]: Yes, in a couple of functions, such as [here](https://github.com/keras-team/keras/blob/master/keras/engine/training.py#L3246). Although they might still work if we completely replaced concat for all of them, I didn't want to risk it.

File: keras/engine/training.py [hunk 2]
```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
   return tf.concat(tensors, axis=axis)
 
 
+def potentially_variable_concat(tensors):
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3261]: Good idea, will replace!

File: keras/engine/training.py [hunk 3]
```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
...
+  """Concats `tensor`s along `axis`. When non-batch dimensions are variable, a RaggedTensor is returned instead."""
+  if isinstance(tensors[0], tf.SparseTensor):
+    return tf.sparse.concat(axis=0, sp_inputs=tensors)
+  non_batch_shapes = tf.stack([tf.shape(tensor)[1:] for tensor in tensors])
+  constant_dims = tf.math.reduce_all(non_batch_shapes == non_batch_shapes[:1], axis=0)
+  if tf.math.reduce_all(constant_dims).numpy().item():  # All non-batch dimensions are constant
+    return tf.concat(tensors, axis=0)
+  # First, identify constant inner dimensions by finding the rightmost dimension that is not constant
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3269]: Because of the `reduce_all`, I believe this conditional is only `True` if all dimensions are constant - the tensor will only have a single element at the point when we call `item()`. Let me know if I overlooked something, though!

[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3269]: Note that `non_batch_shapes` is a 2D tensor of shape `(num_batches, num_non_batch_dims)` - the first `reduce_all` flattens that to `(num_non_batch_dims,)` and the second flattens it to a single element.

File: keras/engine/training.py [hunk 4]
```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
   return tf.concat(tensors, axis=axis)
 
 
+def potentially_variable_concat(tensors):
+  """Concats `tensor`s along `axis`. When non-batch dimensions are variable, a RaggedTensor is returned instead."""
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3262]: Done!

File: keras/engine/training.py [hunk 5]
```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
...
+  if tf.math.reduce_all(constant_dims).numpy().item():  # All non-batch dimensions are constant
+    return tf.concat(tensors, axis=0)
+  # First, identify constant inner dimensions by finding the rightmost dimension that is not constant
+  constant_inner_dimensions = constant_dims.numpy().tolist()[::-1].index(False)
+  # If there are constant inner dimensions, define a constant inner shape
+  if constant_inner_dimensions == 0:
+    constant_inner_shape = None
+  else:
+    constant_inner_shape = tensors[0].shape[-constant_inner_dimensions:]
+  return tf.ragged.constant([tensor.numpy() for tensor in tensors], inner_shape=constant_inner_shape).merge_dims(0, 1)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3283]: I added the switch to `numpy()` because I struggled to make this work any other way - `tf.ragged.constant` does not accept `tf.Tensor` inputs gracefully, and other methods that do like `tf.ragged.stack` often threw errors when the batch dimensions varied.

File: keras/engine/training.py [hunk 6]
```diff
@@ -3258,6 +3258,24 @@ def concat(tensors, axis=0):
   return tf.concat(tensors, axis=axis)
 
 
+def potentially_variable_concat(tensors):
+  """Concats `tensor`s along `axis`. When non-batch dimensions are variable, a RaggedTensor is returned instead."""
+  if isinstance(tensors[0], tf.SparseTensor):
+    return tf.sparse.concat(axis=0, sp_inputs=tensors)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3271]: This is a very good point, and I didn't notice it at all. I don't believe we can create a ragged tensor of sparse tensors, and densifying lots of sparse matrices is very likely to cause OOM errors. The error thrown by `sparse.concat` may be sufficient, but I could also wrap it in a try/except block and raise a more useful error if you want?

File: keras/engine/training.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/engine/training.py:3283]: > tf.ragged.constant does not accept tf.Tensor inputs gracefully

It's really strange that it works with numpy arrays but not tf.Tensors. That's an inconsistency in the ragged API that we should address...

File: keras/engine/training.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/engine/training.py:2014]: This function will be used to concatenate outputs across replicas within `predict` (and `fit` and `evaluate` too), and each replica processes independent batches, which may feature different shapes. So if we don't replace it across the board, `predict` will not work with variable shapes in a distributed setup.

File: keras/engine/training.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/engine/training.py:3271]: > The error thrown by sparse.concat may be sufficient

I suspect it will be impossible to debug for average users. What does it look like today?

File: keras/engine/training.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:2014]: Hm, this is interesting! We commonly use models in multi-GPU `MirroredStrategy()` contexts being fed data from a `tf.data.Dataset` which emits variable-length batches and have not encountered errors until now, but now that you mention it I'm not sure *why* we haven't.

[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:2014]: Possibly when using `fit()`, batches that are passed by the `Dataset` are split across workers after padding has been performed, and therefore those sub-batches are all guaranteed to have the same shape, and the problem does not arise?

File: keras/engine/training.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3283]: I think the cause is that `tf.ragged.constant` is not really intended for this use-case - there are other methods like `tf.RaggedTensor.from_tensor()` that are. However, I couldn't figure out any approach using those methods that preserved constant inner dimensions and didn't crash in edge cases.

File: keras/engine/training.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply keras/engine/training.py:3271]: The current error is something like this:
```
Input shapes must match: expected 128 for dimension 1 but got 126 at position 1 [Op:SparseConcat]
```
This is similar to the error currently thrown by `predict()` when batches have variable sequence length.

**Comment by Rocketknight1** [comment]
Should be ready after tests are re-run!

File: keras/engine/training.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/engine/training.py:3271]: Ok, we can keep the current behavior for the time being. Thanks for checking!

File: keras/tests/integration_test.py [hunk 7]
```diff
@@ -317,5 +317,52 @@ def test_serialization_v2_model(self):
...
+        keras.layers.Conv1D(4, 5, padding='same', activation='relu'),
+        keras.layers.Conv1D(8, 5, padding='same'),
+        keras.layers.BatchNormalization(),
+        keras.layers.Conv1D(3, 5, padding='same', activation='softmax'),
+    ]
+    model = test_utils.get_model_from_layers(
+        layers, input_shape=(None,))
+    model.compile(
+        loss='sparse_categorical_crossentropy',
+        optimizer=keras.optimizers.optimizer_v2.adam.Adam(0.005),
```
Comment (fchollet) [inline keras/tests/integration_test.py:353]: `keras.optimizers.Adam` should work

File: keras/tests/integration_test.py [hunk 8]
```diff
@@ -317,5 +317,52 @@ def test_serialization_v2_model(self):
     loaded_model = keras.models.load_model(output_path)
     self.assertEqual(model.summary(), loaded_model.summary())
 
+@test_combinations.run_with_all_model_types
+@test_combinations.run_all_keras_modes
```
Comment (fchollet) [inline keras/tests/integration_test.py:321]: You may want to avoid running in V1 mode

File: keras/tests/integration_test.py [hunk 9]
```diff
@@ -317,5 +317,52 @@ def test_serialization_v2_model(self):
...
+class TokenClassificationIntegrationTest(test_combinations.TestCase):
+  """Tests a very simple token classification model.
+
+  The main purpose of this test is to verify that everything works as expected when
+  input sequences have variable length, and batches are padded only to the maximum length
+  of each batch. This is very common in NLP, and results in the sequence dimension varying
+  with each batch step for both the features and the labels.
+  """
+  def test_token_classification(self):
+    np.random.seed(1337)
```
Comment (fchollet) [inline keras/tests/integration_test.py:331]: You can improve reproducibility (useful since you check accuracy) with `keras.utils.set_random_seed`

**Review by fchollet — COMMENTED** [review]
Thanks for the updates! Please shorten lines to be within 80 characters.

File: keras/tests/integration_test.py (hunk 8 above)
Comment (Rocketknight1) [reply keras/tests/integration_test.py:321]: Done!

File: keras/tests/integration_test.py (hunk 9 above)
Comment (Rocketknight1) [reply keras/tests/integration_test.py]: Done!

File: keras/tests/integration_test.py (hunk 7 above)
Comment (Rocketknight1) [reply keras/tests/integration_test.py]: Done!

**Comment by fchollet** [comment]
> Please shorten lines to be within 80 characters.

Please do this throughout the code change! Lots of very long lines still :)

**Comment by Rocketknight1** [comment]
Sorry! Should be fixed now.

**Review by fchollet — APPROVED** [review]
LGTM. Thank you for the valuable contribution!

**Comment by Rocketknight1** [comment]
Thanks! I'm happy for it to be merged once you're happy and the CI is green.

**Comment by fchollet** [comment]
Unfortunately I'm seeing some failures in distributed tests, e.g. here:

```
"keras/distribute/distribute_strategy_test.py", [line 1227], in test_fit_eval_and_predict_with_optimizer
    model.predict(get_predict_dataset(distribution), steps=2)
```

The error:

```
"keras/engine/training.py", [line 3279], in potentially_ragged_concat
    non_batch_shapes == non_batch_shapes[:1], axis=0)
...
Invalid reduction dimension (0 for input with 0 dimension(s)
```

Are you able to reproduce this?

**Comment by fchollet** [comment]
Best I can tell, it's likely just a matter of checking for empty arrays and skipping those.

**Comment by Rocketknight1** [comment]
I wasn't able to reproduce that running the tests locally, but that's probably my fault! I've made a change that should fix it, and I'll see if I can get the tests working too.

**Review by fchollet — APPROVED** [review]
Thanks -- let's try this again :)

**Comment by fchollet** [comment]
I've found the origin of the issue, and also why the failure didn't show up on CI: it only occurs with TF 1. We don't run TF 1 tests on CI here on GitHub. I'll fix the issue and merge.

**Comment by Rocketknight1** [comment]
Awesome, thank you!

**Comment by Rocketknight1** [comment]
That's great, thanks for your help throughout!
